bundles/org.openhab.binding.gridbox/src/main/java/org/openhab/binding/gridbox/internal/GridBoxHandler.java

```diff
@@ -0,0 +1,207 @@
+                    new QuantityType<>(evChargingStation.getPower(), Units.WATT));
+        }
+
+        updateState(GridBoxBindingConstants.HEAT_PUMP_POWER, new QuantityType<>(newLiveData.getHeatPump(), Units.WATT));
+        updateState(GridBoxBindingConstants.PHOTOVOLTAIC_PRODUCTION,
+                new QuantityType<>(newLiveData.getPhotovoltaic(), Units.WATT));
+        updateState(GridBoxBindingConstants.PRODUCTION, new QuantityType<>(newLiveData.getProduction(), Units.WATT));
+        updateState(GridBoxBindingConstants.SELF_CONSUMPTION,
+                new QuantityType<>(newLiveData.getSelfConsumption(), Units.WATT));
+        updateState(GridBoxBindingConstants.SELF_CONSUMPTION_RATE,
+                new QuantityType<>(newLiveData.getSelfConsumptionRate() * 100, Units.PERCENT));
+        updateState(GridBoxBindingConstants.SELF_SUFFICIENCY_RATE,
+                new QuantityType<>(newLiveData.getSelfSufficiencyRate() * 100, Units.PERCENT));
+        updateState(GridBoxBindingConstants.SELF_SUPPLY, new QuantityType<>(newLiveData.getSelfSupply(), Units.WATT));
+        updateState(GridBoxBindingConstants.TOTAL_CONSUMPTION,
+                new QuantityType<>(newLiveData.getTotalConsumption(), Units.WATT));
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(GridBoxConfiguration.class);
+        if (config.email == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.configuration-error.noemail");
+            return;
+        }
+        if (config.password == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.configuration-error.nopassword");
+            return;
+        }
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduler.execute(this::initializeApi);
+    }
+
+    private void initializeApi() {
+        updateStatus(ThingStatus.UNKNOWN);
+
+        try {
+            config.idToken = API.getIdToken(config);
+
+            String systemId = config.systemId;
+            if (systemId == null || systemId.isBlank()) {
+                config.systemId = API.getSystemId(config);
+            }
+            updateStatus(ThingStatus.ONLINE);
+
+            updateScheduledFuture = scheduler.scheduleWithFixedDelay(this::update, 0, config.refreshInterval,
+                    TimeUnit.SECONDS);
+        } catch (GridBoxApiAutheticationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.configuration-error.credentialsinvalid");
+        } catch (IOException | InterruptedException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "@test/offline.communication-error.connectionlost");
+            scheduler.schedule(this::initialize, CONNECTION_RETRY_PERIOD, TimeUnit.SECONDS);
```

Comment:
Also add timeoutexception here

Reply:
Please have a look at the new implementation, did you mean something like that?

Reply:
Yeah, better now. There are some path's that i'm a bit uncertain about. If `GridBoxApiAutheticationException` is thrown, it immediatley re-initializes and the connection attemps is reset to 0. If something is wrong, it can immediate throw the `GridBoxApiAutheticationException` again causing some kind of loop without backoff or fatal error. 
Maybe only reset the reconnectattempt to 0 when a succecfull API call has been made. And additional give up after 10 retries and mark the thing offline ?

Reply:
I also added some unit tests for the API, maybe you can have a look at them?

Reply:
Great thanks!